x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/RemoteAccessAuthenticationService.java

```diff
@@ -0,0 +1,252 @@
+
+        // TODO revisit this once Node's Version is refactored
+        if (getMinNodeVersion().before(Authentication.VERSION_REMOTE_ACCESS_REALM)) {
+            withRequestProcessingFailure(
+                authcContext,
+                new IllegalArgumentException(
+                    "all nodes must have version ["
+                        + Authentication.VERSION_REMOTE_ACCESS_REALM
+                        + "] or higher to support cross cluster requests with remote access"
+                ),
+                listener
+            );
+            return;
+        } else if (threadContext.getHeader(AuthenticationField.AUTHENTICATION_KEY) != null) {
+            withRequestProcessingFailure(
+                authcContext,
+                new IllegalArgumentException("authentication header is not allowed with remote access"),
+                listener
+            );
+            return;
+        } else if (threadContext.getHeader(AuthenticationServiceField.RUN_AS_USER_HEADER) != null) {
+            withRequestProcessingFailure(
+                authcContext,
+                new IllegalArgumentException("run-as header is not allowed with remote access"),
+                listener
+            );
+            return;
+        }
+
+        final RemoteAccessHeaders remoteAccessHeaders;
+        try {
+            remoteAccessHeaders = extractRemoteAccessHeaders(threadContext);
+        } catch (Exception ex) {
+            withRequestProcessingFailure(authcContext, ex, listener);
+            return;
+        }
+
+        try (
+            ThreadContext.StoredContext ignored = threadContext.newStoredContext(
+                Collections.emptyList(),
+                // drop remote access authentication headers since we've read their values, and we want to maintain the invariant that
+                // either the remote access authentication header is in the context, or the authentication header, but not both
+                List.of(REMOTE_ACCESS_CLUSTER_CREDENTIAL_HEADER_KEY, REMOTE_ACCESS_AUTHENTICATION_HEADER_KEY)
+            )
+        ) {
+            final Supplier<ThreadContext.StoredContext> storedContextSupplier = threadContext.newRestorableContext(false);
+            authcContext.addAuthenticationToken(remoteAccessHeaders.clusterCredential());
+            authenticationService.authenticate(
+                authcContext,
+                new ContextPreservingActionListener<>(storedContextSupplier, ActionListener.wrap(authentication -> {
+                    assert authentication.isApiKey() : "initial authentication for remote access must be by API key";
+                    assert false == authentication.isRunAs() : "initial authentication for remote access cannot be run-as";
+                    final RemoteAccessAuthentication remoteAccessAuthentication = remoteAccessHeaders.remoteAccessAuthentication();
+                    validate(remoteAccessAuthentication);
```

Comment:
Validating here for two reasons: I don't think we should validate role descriptors before we've established that the request issuer has valid credentials. I see this getting abused, e.g. someone without an API key submitting large or bogus role descriptors. Secondly, once we do the refactoring to move validation further down to role building this will happen after authentication either way.

Reply:
👍